File: Numerical_Methods/Numerical Integration Techniques/Simpson’s 13 Rule/java_solutions/Problem9.java
import java.util.Scanner;

public class Problem9 {
    // Function to calculate ρ(x) = x^2
    public static double density(double x) {
        return x * x;
    }

    // Function to calculate x^2 * ρ(x) for moment of inertia
    public static double xSquaredDensity(double x) {
        return x * x * density(x);
    }

    // Simpson's 1/3 Rule implementation
    public static double simpsonsRule(double a, double b, int n, boolean isXSquaredDensity) {
        if (n % 2 != 0) {
            throw new IllegalArgumentException("Number of intervals must be even");
        }

        double h = (b - a) / n;
        double sum = isXSquaredDensity ? xSquaredDensity(a) + xSquaredDensity(b) : density(a) + density(b);

        for (int i = 1; i < n; i++) {
            double x = a + i * h;
            sum += (i % 2 == 0) ? 
                2 * (isXSquaredDensity ? xSquaredDensity(x) : density(x)) : 
                4 * (isXSquaredDensity ? xSquaredDensity(x) : density(x));
        }

        return (h / 3) * sum;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        
        try {
            // Fixed values for this problem
            double a = 0.0;
            double b = 1.0;
            int n = 6;

            // Calculate total mass
            double totalMass = simpsonsRule(a, b, n, false);
            
            // Calculate moment of inertia
            double momentOfInertia = simpsonsRule(a, b, n, true);
            
            // Exact value of moment of inertia
            // For ρ(x) = x^2, the exact moment of inertia is 1/6
            double exactValue = 1.0 / 6.0;
            
            // Calculate error
            double error = Math.abs(momentOfInertia - exactValue);

            // Print results
            System.out.println("Moment of Inertia Calculation using Simpson's 1/3 Rule:");
            System.out.println("----------------------------------------------------");
            System.out.printf("Number of intervals (n): %d%n", n);
            System.out.printf("Total mass: %.10f%n", totalMass);
            System.out.printf("Approximate moment of inertia: %.10f%n", momentOfInertia);
            System.out.printf("Exact moment of inertia: %.10f%n", exactValue);
            System.out.printf("Absolute error: %.10f%n", error);
            System.out.printf("Relative error: %.10f%%%n", (error / exactValue) * 100);

            // Show how accuracy changes with different values of n
            System.out.println("\nAccuracy Analysis with Different n Values:");
            System.out.println("----------------------------------------");
            for (int i = 2; i <= 20; i += 2) {
                double approx = simpsonsRule(a, b, i, true);
                double err = Math.abs(approx - exactValue);
                System.out.printf("n = %2d: Error = %.10f%n", i, err);
            }

        } catch (Exception e) {
            System.out.println("Error: " + e.getMessage());
        } finally {
            scanner.close();
        }
    }
} 

File: Numerical_Methods/Numerical Integration Techniques/Simpson’s 13 Rule/python_solutions/problem9.py
import numpy as np

def density(x):
    """Function to calculate ρ(x) = x^2"""
    return x**2

def x_squared_density(x):
    """Function to calculate x^2 * ρ(x) for moment of inertia"""
    return x**2 * density(x)

def simpsons_rule(a, b, n, is_x_squared_density=False):
    """
    Implementation of Simpson's 1/3 Rule
    
    Parameters:
    a : float - Lower limit of integration
    b : float - Upper limit of integration
    n : int - Number of intervals (must be even)
    is_x_squared_density : bool - Whether to use x_squared_density function
    
    Returns:
    float - Approximate value of the integral
    """
    if n % 2 != 0:
        raise ValueError("Number of intervals must be even")
    
    h = (b - a) / n
    x = np.linspace(a, b, n+1)
    y = x_squared_density(x) if is_x_squared_density else density(x)
    
    # Apply Simpson's 1/3 rule formula
    result = h/3 * (y[0] + y[-1] + 4*np.sum(y[1:-1:2]) + 2*np.sum(y[2:-1:2]))
    return result

def main():
    try:
        # Fixed values for this problem
        a = 0.0
        b = 1.0
        n = 6
        
        # Calculate total mass
        total_mass = simpsons_rule(a, b, n)
        
        # Calculate moment of inertia
        moment_of_inertia = simpsons_rule(a, b, n, is_x_squared_density=True)
        
        # Exact value of moment of inertia
        # For ρ(x) = x^2, the exact moment of inertia is 1/6
        exact_value = 1/6
        
        # Calculate error
        error = abs(moment_of_inertia - exact_value)
        
        # Print results
        print("Moment of Inertia Calculation using Simpson's 1/3 Rule:")
        print("----------------------------------------------------")
        print(f"Number of intervals (n): {n}")
        print(f"Total mass: {total_mass:.10f}")
        print(f"Approximate moment of inertia: {moment_of_inertia:.10f}")
        print(f"Exact moment of inertia: {exact_value:.10f}")
        print(f"Absolute error: {error:.10f}")
        print(f"Relative error: {(error/exact_value)*100:.10f}%")
        
        # Show how accuracy changes with different values of n
        print("\nAccuracy Analysis with Different n Values:")
        print("----------------------------------------")
        for i in range(2, 21, 2):
            approx = simpsons_rule(a, b, i, is_x_squared_density=True)
            err = abs(approx - exact_value)
            print(f"n = {i:2d}: Error = {err:.10f}")
        
    except Exception as e:
        print(f"Error: {str(e)}")

if __name__ == "__main__":
    main() 